### ISSUE TITLE
ingress rules gets random order / priority in load balancer listener rules

### ISSUE THREAD

**Issue by ThomasSteinbach** [kubernetes-sigs/aws-load-balancer-controller#2203]
/kind bug

**Describe the bug**

The aws-load-balancer-controller creates Load Balancer Listener Rules in the same order as K8s ingress rules are defined in the ingress object. This does NOT correctly implement the K8s ingress specification, as the ingress documentation says:

> In some cases, multiple paths within an Ingress will match a request. In those cases precedence will be given first to the longest matching path.
( https://kubernetes.io/docs/concepts/services-networking/ingress/#multiple-matches )

More worse - the order of rules within an ingress specification is not ensured, as those rules are just a YAML/JSON list with no defined order. When processing the specification, any tool could possibly "sort" the items or alter the order. This is for instance the case if you process your k8s yaml specifications with official kubernetes kustomize.

That means that the priority of load balancer listener is random. At least NOT ordered by the longest matching path.

**Steps to reproduce**

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
  name: my-service
spec:
  rules:
  - host: my-service.hello.world
    http:
      paths:
      - backend:
          service:
            name: my-service
            port:
              name: application
        path: /
        pathType: Prefix
      - backend:
          service:
            name: my-service
            port:
              name: monitoring
        path: /actuator/
        pathType: Prefix
```

The AWS Load Balancer created will forward requests always to the `application` port. Even in the case of the request path `/actuator/` he will not forward to the `monitoring` port, as the rule with the path `/` get a higher priority in the Load Balancer Listener Rules.

**Expected outcome**

* The listener rule with the more specific path `/actuator` should get a higher priority in the Load Balancer Listener Rules.
* For the upper example requests to `my-service.hello.world/actuator` should be forwarded to the `monitoring` port.

**Environment**

* AWS Load Balancer controller version: helm chart versiono 1.2.3

**Comment by kishorj** [comment]
@ThomasSteinbach, do the listener rules follow the order you specify in the manifest?

**Comment by ThomasSteinbach** [comment]
@kishorj yes, they do.

**Comment by kishorj** [comment]
/kind bug
we will fix it in the next release

**Comment by sergeyshevch** [comment]
Also same behavior if we have multiple kind:Ingress with the same domain and different paths:
Example:

Ingress a contains host example.com and path /
Ingress b contains host example.com and path /blog/

**Comment by sergeyshevch** [comment]
@oliviassss Is there any estimations when it can be fixed?

**Comment by benjamin658** [comment]
Could we manually reorder the rule by AWS console before the bug fixing?

**Comment by oliviassss** [comment]
@benjamin658 I think it's an option. You can go to the AWS console to select the specific load balancer, then select the `Listener` tab, and click on the `view/edit rules`. Then you can see all the rules for this ingress. There's an arrow icon on the menu bar, which is to reorder the rules.
![Screen Shot 2022-01-31 at 12 07 15 PM](https://user-images.githubusercontent.com/86027005/151865416-a4777bc9-c785-4f8b-9f5a-a2ddb9aa95c7.png)

**Comment by benjamin658** [comment]
@oliviassss , Thanks for your explanation.

We deploy multiple ingresses manifest in the same group by helm chart, it's hard to order each manifest's rules appropriately before deploying. 

Currently, we change our route paths to avoid the same prefix.

**Comment by ngoyal16** [comment]
+1

**Comment by sergeyshevch** [comment]
I guess that #2409 is related to this issue and will close it

**Comment by ngoyal16** [comment]
@sergeyshevch when it is plan to release any estimates?

**Comment by sergeyshevch** [comment]
@ngoyal16 Sorry but it's not my PR so I guess you need to address this question to @oliviassss

**Comment by kishorj** [comment]
@sergeyshevch, we held merging the fix for prior release due to concerns on backwards compatibility. We are working on it.

**Comment by ngoyal16** [comment]
@kishorj any ETA on same?

**Comment by kishorj** [comment]
We will include the fix in the v2.4.2 patch release.

**Comment by ngoyal16** [comment]
@kishorj  any ETA for v2.4.2 patch release

**Comment by ngoyal16** [comment]
@kishorj @oliviassss  is it live? if not any ETA.. it is much awaiting feature.

**Comment by roumigus** [comment]
As far as I can tell (by checking this repo), this is not live. https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/2409 fixes the issue but has been merged after the release of 2.4.2

We're thus waiting for a release including the changes in the merged PR above.

It is a serious bug in my opinion. Using `prefix` in production without understanding this issue can be dangerous, depending on the case. Thank you @ThomasSteinbach for the clear bug report!

> @sergeyshevch, we held merging the fix for prior release due to concerns on backwards compatibility. We are working on it.

For reference, here is how backward compatibility have been handled (in case anyone want to predict the future behavior): https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/2409#issuecomment-1083375194

**Comment by kishorj** [comment]
Fix available in v2.4.3 release, closing the issue
